src/udp/connection/encoded_id.rs

```diff
@@ -0,0 +1,346 @@
+        let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+        let witness = hasher.finish().to_le_bytes();
+
+        let id: Vec<u8> = [&witness[0..4], expiry_u32.to_le_bytes().as_slice()].concat();
+
+        WitnessEncodedId {
+            value: id.try_into().unwrap(),
+        }
+    }
+
+    fn value(&self) -> &[u8; 8] {
+        &self.value
+    }
+
+    fn check(
+        client_id: KeyedClientId,
+        _lifetime: Duration,
+        encoded_id: [u8; 8],
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        let expiry_bytes: [u8; 4] = encoded_id[4..8].try_into().unwrap();
+        let expiry_u32 = u32::from_le_bytes(expiry_bytes);
+
+        let after_maybe = match after_now {
+            Some(after) => after,
+            None => Duration::ZERO,
+        };
+
+        let now_u32 = LocalClock::next_as_timestamp(&after_maybe).try_into().unwrap();
+
+        if expiry_u32 <= now_u32 {
+            return Err("Expired connection id");
+        }
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+
+        let witness = hasher.finish().to_le_bytes();
+
+        if &witness[0..4] != &encoded_id[0..4] {
+            Err("Bad Witness in Connection Id")
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl EncodedId<PlainClientId> for EncryptedEncodedId {
+    type Error = &'static str;
+
+    fn new(client_id: PlainClientId, lifetime: Duration) -> Self {
+        let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
+
+        let id_clear: Vec<u8> = [&client_id.value()[0..4], &expiry_u32.to_le_bytes().as_slice()].concat();
+
+        let id_clear_bytes: [u8; 8] = id_clear.try_into().unwrap();
+        let mut block: BlowfishArray = BlowfishArray::from(id_clear_bytes);
+
+        let blowfish_key: &blowfish::Blowfish<LE> = if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };
+
+        <Blowfish<LE> as BlockEncrypt>::encrypt_block(blowfish_key, &mut block);
+
+        EncryptedEncodedId {
+            value: block.try_into().unwrap(),
+        }
+    }
+
+    fn value(&self) -> &[u8; 8] {
+        &self.value
+    }
+
+    fn check(
+        client_id: PlainClientId,
+        _lifetime: Duration,
+        encoded_id: [u8; 8],
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        let mut block: BlowfishArray = BlowfishArray::from(encoded_id);
+
+        let blowfish_key: &blowfish::Blowfish<LE> = if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };
+
+        <Blowfish<LE> as BlockDecrypt>::decrypt_block(&blowfish_key, &mut block);
+
+        let client_id_bytes: [u8; 4] = block[0..4].try_into().unwrap();
```

Comment:
@da2ce7 I would rename this to:

```rust
let client_id_witness_bytes: [u8; 4] = block[0..4].try_into().unwrap();
```

and also where you use it:

```rust
let expected_client_id_witness_bytes = client_id.value()[0..4];
if &client_id_witness_bytes != &client_id.value()[0..4] {
    Err("Bad Witness in Connection Id")
} else {
    Ok(())
}
```

I would make it more explicit that we are using only 4 bytes from the client id to check that the connection id is not being used by other clients.